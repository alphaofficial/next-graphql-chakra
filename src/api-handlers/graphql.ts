import HASURA_CLIENT from "@/api-lib/client";
import { NextApiRequest, NextApiResponse } from "next";

export async function graphqlHandler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  try {
    const response = await HASURA_CLIENT.post("", {
      query: req.body.query,
      variables: req.body.variables,
    });
    if (response.data) {
      res.status(200).json(response.data);
      return;
    }
    res.status(500).json({ data: null });
  } catch (error) {
    console.log({ error });
    res.status(500).json({ data: null });
  }
}
