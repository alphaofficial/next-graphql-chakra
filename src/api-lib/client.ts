import axios from "axios";

const { HASURA_GRAPHQL_ADMIN_SECRET, NEXT_PUBLIC_HASURA_GRAPHQL_URL } =
  process.env;

const HASURA_CLIENT = axios.create({
  baseURL: NEXT_PUBLIC_HASURA_GRAPHQL_URL,
  headers: {
    "x-hasura-admin-secret": HASURA_GRAPHQL_ADMIN_SECRET,
  },
});

export default HASURA_CLIENT;
