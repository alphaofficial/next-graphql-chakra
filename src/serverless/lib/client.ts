// Seems to be a problem with graphql-request 5.2.0, pinning it to 5.1.0 fixes the problem
import { GraphQLClient } from "graphql-request";
import { getSdk } from "@/serverless/generated/server-sdk";
import config from "@/common/config";

export const graphQLClient = new GraphQLClient(config.GRAPHQL_URL, {
  headers: {
    "x-hasura-admin-secret": config.GRAPHQL_ADMIN_SECRET,
  },
});

const graphqlSDK = getSdk(graphQLClient);
export default graphqlSDK;
