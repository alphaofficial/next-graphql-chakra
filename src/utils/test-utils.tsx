// test-utils.js
import { render, RenderOptions } from "@testing-library/react";
import * as React from "react";
import userEvent from "@testing-library/user-event";
import {
  Hydrate,
  QueryClient,
  QueryClientProvider,
} from "@tanstack/react-query";
import { ChakraProvider } from "@chakra-ui/react";

const user = userEvent.setup();
const queryClient = new QueryClient();
const AllTheProviders: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  return (
    <QueryClientProvider client={queryClient}>
      <Hydrate state={{}}>
        <ChakraProvider>{children}</ChakraProvider>
      </Hydrate>
    </QueryClientProvider>
  );
};

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "queries">
) =>
  render(ui, {
    wrapper: AllTheProviders as React.ComponentType,
    ...options,
  });

// re-export everything
export * from "@testing-library/react";
// override render method
export { customRender as render, user };

export async function addRecordingBufferTime(timeout?: number): Promise<any> {
  if (process.env.SHOULD_RECORD_VIDEO === "true") {
    return new Promise((resolve) =>
      setTimeout(() => resolve(true), timeout || 5000)
    );
  }
  return Promise.resolve(true);
}
