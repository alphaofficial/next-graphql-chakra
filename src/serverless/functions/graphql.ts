import { NextApiRequest, NextApiResponse } from "next";
import { graphQLClient } from "@/serverless/lib/client";

export async function graphqlHandler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  try {
    const { query, variables } = req.body;
    const data = await graphQLClient.request(query, variables);
    res.status(200).json({ data });
  } catch (error) {
    console.log({ error });
    res.status(500).json({ data: null });
  }
}
