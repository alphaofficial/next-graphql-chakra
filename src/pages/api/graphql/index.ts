import { graphqlHandler } from "@/serverless/functions/graphql";
import Handler from "@/serverless/helpers/handler";

/**
 * This is the handler for the /api/graphql
 */
const api = new Handler();
const gql = api.handler.post(graphqlHandler);

export default gql;
