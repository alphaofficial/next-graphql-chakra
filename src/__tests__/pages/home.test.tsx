import { CharactersByVillageQuery } from "@/client/generated/graphql";
import { render, waitFor } from "@/client/test-utils";
import Home from "@/pages/index";

const mockData = {
  characters: [
    {
      id: "1",
      name: "Naruto Uzumaki",
      avatarSrc: "https://via.placeholder.com/150",
    },
    {
      id: "2",
      name: "Sasuke Uchiha",
      avatarSrc: "https://via.placeholder.com/150",
    },
  ],
} as CharactersByVillageQuery;

const props = {
  data: mockData,
  dehydratedState: {},
};
const renderPage = () => render(<Home {...props} />);
describe("Home page", () => {
  jest.mock("@tanstack/react-query", () => {
    const originalModule = jest.requireActual("@tanstack/react-query");
    return {
      __esModule: true,
      ...originalModule,
      useCharactersByVillageQuery: jest.fn(),
    };
  });

  it("renders correctly", async () => {
    const { container } = renderPage();
    const node = await waitFor(() => container);
    expect(node).toMatchSnapshot();
  });
});
