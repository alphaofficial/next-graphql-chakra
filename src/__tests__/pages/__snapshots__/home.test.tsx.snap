// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page renders correctly 1`] = `
<div>
  <div
    class="css-0"
  >
    <div
      class="css-1glkkdp"
    >
      <div
        class="css-0"
      >
        <h2
          aria-label="heading"
          class="chakra-heading css-1dklj6k"
        >
          Hajimemashite 👋🏽!
        </h2>
      </div>
      <div
        class="chakra-stack css-ojjdrs"
      >
        <h2
          aria-label="heading"
          class="chakra-heading css-18j379d"
        >
          Characters
        </h2>
        <div
          class="chakra-stack css-ojjdrs"
        >
          <p
            class="chakra-text css-0"
          >
            No users found
          </p>
          <div
            class="css-15pd5yp"
          />
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
